src/Microsoft.ServiceFabric.Actors/Runtime/MigrationOrchestrator.cs

```diff
@@ -0,0 +1,452 @@
+                        cancellationToken.ThrowIfCancellationRequested();
+                        await this.StartDowntimePhaseWorkload(lastAppliedSN, cancellationToken);
+                    }
+                    else
+                    {
+                        int.TryParse(Encoding.ASCII.GetString(iterationValue.Value), out int iterationCount);
+                        ConditionalValue<byte[]> lastSNValue, lastAppliedSNValue;
+                        using (var tx = this.stateProvider.GetStateManager().CreateTransaction())
+                        {
+                            lastSNValue = await this.metadataDict.TryGetValueAsync(tx, MigrationConstants.GetCatchupWorkerEndSNKey(iterationCount));
+                            lastAppliedSNValue = await this.metadataDict.TryGetValueAsync(tx, MigrationConstants.GetCatchupWorkerLastAppliedSNKey(iterationCount));
+                        }
+
+                        if (lastAppliedSNValue.HasValue)
+                        {
+                            long.TryParse(Encoding.ASCII.GetString(lastAppliedSNValue.Value), out long lastAppliedSN);
+                            long.TryParse(Encoding.ASCII.GetString(lastSNValue.Value), out long lastSN);
+                            var endSequenceNumber = long.Parse(await Client.GetStringAsync(this.kvsEndpoint + MigrationConstants.GetEndSNEndpoint));
+
+                            if (lastAppliedSN != lastSN || (endSequenceNumber - lastAppliedSN > this.downtimeThreshold))
+                            {
+                                lastAppliedSN = await this.StartCatchupIteration(iterationCount, lastAppliedSN, cancellationToken);
+                                await this.StartDowntimePhaseWorkload(lastAppliedSN, cancellationToken);
+                            }
+                            else
+                            {
+                                cancellationToken.ThrowIfCancellationRequested();
+                                await this.StartDowntimePhaseWorkload(lastAppliedSN, cancellationToken);
+                            }
+                        }
+                        else
+                        {
+                            ConditionalValue<byte[]> prevIterationLastAppliedSNValue;
+                            using (var tx = this.stateProvider.GetStateManager().CreateTransaction())
+                            {
+                                prevIterationLastAppliedSNValue = await this.metadataDict.TryGetValueAsync(tx, MigrationConstants.GetCatchupWorkerLastAppliedSNKey(iterationCount - 1));
+                            }
+
+                            long.TryParse(Encoding.ASCII.GetString(prevIterationLastAppliedSNValue.Value), out long prevIterationLastAppliedSN);
+                            var lastAppliedSN = await this.StartCatchupIteration(iterationCount - 1, prevIterationLastAppliedSN, cancellationToken);
```

Comment:
Should we not start the new iteration? why are you starting the phase with iterationcount -1